Social Sign Up: Warn the user about an already existing account

zaguiini opened the pull request:
#### Proposed Changes

We want to display a message when the user tries to socially authenticate using an e-mail that is already registered in WPCOM. That already happens in login, but it's not the case in the signup flow.

Well, it "kind of" happens: we're redirecting the user immediately, not giving them the opportunity to try again and select a different profile. And that process is buggy as (1) it displays the domains step for about a second and (2) it's not setting the e-mail, rendering a blank page so the user needs to re-type it.

That's subpar and we thought of a better experience: tell the user of the problem and give them a link so they could try it _if they want to_. That's better because it respects user preferences.

#### Testing Instructions

1. See "Testing Instructions" in https://github.com/Automattic/wp-calypso/pull/64957 to set up a working wpcalypso.wordpress.com environment on your local machine;
2. Create a WPCOM account with a Google e-mail with the classic e-mail and password method (NOT socially authenticating);

Open a new window, go to https://wpcalypso.wordpress.com/start/user and click **Continue with Google** choosing the same e-mail you've used to create the account. After choosing the account in the pop-up, you should see this screen: 

<img width="1241" alt="image" src="https://user-images.githubusercontent.com/36432/180775890-d200e552-a0e0-491f-b611-430bebe12086.png">

Clicking on the link should redirect you to the login page, with the e-mail being prefilled and form representing the social linking state.

<img width="1240" alt="image" src="https://user-images.githubusercontent.com/36432/180791852-cf7f0836-8f7b-4c8a-940a-c7c063a838cc.png">

Now, open https://wordpress.com/me/security/social-login and socially connect with Google there. After that's done, when you try to sign in with Google at https://wpcalypso.wordpress.com/start/user, you should be redirected to `/start/domains` instead of being alerted about an existing user.

Related to https://github.com/Automattic/wp-calypso/issues/65654.

zaguiini committed:
Only go to next step when the previous one is completed

zaguiini committed:
Allow disabling redirecting if social connect error happens

zaguiini committed:
Handle user_exists error on sign up component

zaguiini committed:
Rename user_exists error display prop to something more descriptive

zaguiini committed:
Improve copy

danielbachhuber committed:
Merge branch 'trunk' into improve/warn-user-existing-account-signup

danielbachhuber committed:
Tweak language for greater clarity on signup form

danielbachhuber reviewed (commented):
This works as documented. I made a slight tweak to the message, to reflect its display on the signup page.

One thing I noticed, though.

~~Instead, I think we should take them to this page when the click on the "Log in to this account" link:~~ Implemented!

danielbachhuber committed:
Merge branch 'trunk' into improve/warn-user-existing-account-signup

danielbachhuber commented:
> I _think_ what's happening is that the global `state` is dumped when `page()` is called.

As it turns out, this is the case!

https://github.com/Automattic/wp-calypso/blob/9fbce67d377a07dfbfbed8b7d194a5abb82d650b/client/landing/login/index.js#L20-L44

danielbachhuber committed:
Social Auth: Hand off the social connect state to `/login` (#65940)

danielbachhuber commented on client/login/controller.js:
```diff
@@ -16,6 +17,25 @@ const enhanceContextWithLogin = ( context ) => {
 
+	// Process a social login handoff from /start/user.
+	if ( query?.email_address && query?.source && query?.access_token && query?.id_token ) {
+		context.store.dispatch( {
```
@p-jackson Oddly, when I'm running this code at `http://calypso.localhost:3000/`, `context.store.dispatch()` successfully sets the store. However, when I'm running the code at `https://wpcalypso.wordpress.com/`, it doesn't set the store. Any ideas why this might be the case?

p-jackson replied:
Is it just a typo thing? During testing the debugger wasn't stepping into the branch of the code that dispatched the action. But I changed `query?.source` to `query?.service` and it worked, and I think `service` is what we want anyway.

danielbachhuber replied:
> Is it just a typo thing? During testing the debugger wasn't stepping into the branch of the code that dispatched the action. But I changed `query?.source` to `query?.service` and it worked, and I think `service` is what we want anyway.

D'oh!

danielbachhuber commented on client/login/controller.js:
```diff
@@ -16,6 +17,25 @@ const enhanceContextWithLogin = ( context ) => {
+		params.delete( 'source' );
+		params.delete( 'access_token' );
+		params.delete( 'id_token' );
+		page.redirect( window.location.pathname + '?' + params.toString() );
```
Any ideas why `page.redirect()` adds a new history entry, instead of replacing the existing?

p-jackson replied:
I think it might be because it's running on the server-side, so `page.js` is doing the real redirect. I was getting server-side errors because it was trying to interact with `window.location` during server-side routing. I've made a change so that it only removes the query params on the client (it's sort of like an optional tidy-up I guess) and it's not adding an extra history entry anymore.

danielbachhuber commented:
@Automattic/team-calypso Could I get an initial review on this pull request? It's mostly functional, except for the quirks identified in the two comments above.

p-jackson committed:
Social signup-to-login handoff requires `service` query param

The handoff code was expecting `source` but that appears to be a typo,
since the link we provide from the signup page to the login page doesn't
set a `source` query param.

p-jackson committed:
Only client removes sensitive query params after social login handoff

p-jackson commented:
@danielbachhuber I believe I've addressed the two issues you mentioned above in bb4feb8 and 78ae762

danielbachhuber commented:
> I believe I've addressed the two issues you mentioned above in [bb4feb8](https://github.com/Automattic/wp-calypso/commit/bb4feb85d44497bbdb694b68ba87fcd53388ac48) and [78ae762](https://github.com/Automattic/wp-calypso/commit/78ae76221e45573d24ca2c633d458b302b89f7bc)

@p-jackson Thank you!!!

noahtallen commented on client/blocks/signup-form/index.jsx:
```diff
@@ -194,9 +196,16 @@ class SignupForm extends Component {
 	 * If the step is invalid because we had an error that the user exists,
 	 * we should prompt user with a request to connect his social account
-	 * to his existing WPCOM account
+	 * to his existing WPCOM account.
```
Nit: "their" would be a better pronoun to use here rather than "his" :)

danielbachhuber replied:
> Nit: "their" would be a better pronoun to use here rather than "his" :)

@noahtallen Changed with b0dadf65de

noahtallen commented on client/blocks/signup-form/index.jsx:
```diff
@@ -194,9 +196,16 @@ class SignupForm extends Component {
+	 * That can be done either by redirecting or only suggesting. If it's done
+	 * by suggesting, bail out of redirecting and display the error.
 	 */
 	maybeRedirectToSocialConnect() {
```
With the new behavior, will we ever redirect to social connect? I might misunderstand, but it seem like the two options now are "create account with different method" or "log in with the same email," but not social connection. 

If that's the case, maybe we can remove this code?

danielbachhuber replied:
> With the new behavior, will we ever redirect to social connect? I might misunderstand, but it seem like the two options now are "create account with different method" or "log in with the same email," but not social connection.
> 
> If that's the case, maybe we can remove this code?

@noahtallen I'm not sure, to be honest ☹️ 

@p-jackson Any idea? Or do we need to wait until @zaguiini is back?

danielbachhuber replied:
> > With the new behavior, will we ever redirect to social connect? I might misunderstand, but it seem like the two options now are "create account with different method" or "log in with the same email," but not social connection.
> > If that's the case, maybe we can remove this code?
> 
> I'm not sure, to be honest ☹️

Oh. It looks like there are other uses of `<SignupForm>`, so we want to maintain backwards compatibility with those.

noahtallen replied:
Cool! Is it easy for this new behavior to apply to those other uses of SignupForm, or are those isolated enough that it wouldn't work?

danielbachhuber replied:
> Is it easy for this new behavior to apply to those other uses of SignupForm, or are those isolated enough that it wouldn't work?

@noahtallen I _think_ so, but I'm not confident it's expected. We've already run outside the initial scope of the project, so I'm not sure we want to spend more time on it.

noahtallen replied:
Fair enough! :)

noahtallen commented on client/login/controller.js:
```diff
@@ -14,8 +15,31 @@ const enhanceContextWithLogin = ( context ) => {
 	} = context;
 
+	// Process a social login handoff from /start/user.
+	if ( query?.email_address && query?.service && query?.access_token && query?.id_token ) {
```
I'm not sure if this is a fair concern, but are there any security concerns putting the access token in the URL?

danielbachhuber replied:
> I'm not sure if this is a fair concern, but are there any security concerns putting the access token in the URL?

@noahtallen It's a great concern!

The access token and id token are: 1) passed along in another URL, and 2) stored in local storage. As such, I think it's fine to expose them again in a URL, although I agree it's less than ideal.

noahtallen replied:
Cool. That makes sense to me 👍

noahtallen commented on client/signup/steps/user/index.jsx:
```diff
@@ -102,6 +102,11 @@ export class UserStep extends Component {
 	};
 
 	componentDidUpdate() {
+		if ( this.props.step?.status === 'completed' ) {
```
I wonder why this wasn't here before! Is this what was causing the one step to flash for a second during the flow?

zaguiini replied:
It was 👍

noahtallen commented on client/signup/steps/user/index.jsx:
```diff
@@ -455,6 +463,7 @@ export class UserStep extends Component {
 					recaptchaClientId={ this.state.recaptchaClientId }
 					horizontal={ isReskinned }
 					isReskinned={ isReskinned }
+					shouldDisplayUserExistsError
```
Is there any time we _don't_ want this behavior?

danielbachhuber replied:
> Is there any time we _don't_ want this behavior?

Per conversation in https://github.com/Automattic/wp-calypso/pull/65908#discussion_r934743942, we always want the error to display in this context. It's a prop because we may or may not want to display the error in other contexts.

noahtallen commented on client/login/controller.js:
```diff
@@ -14,8 +15,31 @@ const enhanceContextWithLogin = ( context ) => {
+	// Process a social login handoff from /start/user.
+	if ( query?.email_address && query?.service && query?.access_token && query?.id_token ) {
+		context.store.dispatch( {
+			type: SOCIAL_HANDOFF_CONNECT_ACCOUNT,
```
Nice! Does this reducer automatically connect the social account when this data exists?

danielbachhuber replied:
> Nice! Does this reducer automatically connect the social account when this data exists?

@noahtallen It doesn't automatically connect, but takes them to this state:

![image](https://user-images.githubusercontent.com/36432/182207320-7df812be-5e55-4947-a819-6213df5302aa.png)

noahtallen replied:
Gotcha. What does the auth info help us do in this case? The way I read the original code, it seemed like we were putting the google auth info into the reducer. I think seeing "social" in the action type made me think that :)

danielbachhuber replied:
> What does the auth info help us do in this case?

@noahtallen The `access_token` and `id_token` are required to complete the connection process in the form above.

> The way I read the original code, it seemed like we were putting the google auth info into the reducer. I think seeing "social" in the action type made me think that :)

I'm not sure I follow... could you explain further?

noahtallen replied:
Ah, I think I misunderstood your initial comment -- it doesn't automatically connect at first, but it does still connect the email/password account to the Google account after you submit the form. That makes a lot of sense and matches what I'd expect as well.

When I read your initial comment, I thought that it wouldn't connect the Google account after logging in with email/pass.

noahtallen reviewed (commented):
Very cool, thanks for working on this! I don't know as much as many Calypso veterans, but hopefully my review is still useful! I mostly asked a few clarifying questions. No major concerns from me.

danielbachhuber committed:
Merge branch 'trunk' into improve/warn-user-existing-account-signup

danielbachhuber committed:
Replace 'his' with 'their'

danielbachhuber committed:
Merge branch 'trunk' into improve/warn-user-existing-account-signup

danielbachhuber committed:
Fire a Tracks event when they click this link

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7462644

Thank you @zaguiini for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7462644).
